### PR TITLE
[LA-7] [LA-8] [LA-59] [LA-73] [LA-17] [LA-79] LTI launch for CloudLRS, UI fixes for Privacy Dash, Ebextension updates. 

### PR DESCRIPTION
**Pull request by sandeep-jay**
The commit contains the following : 
1) LTI provider for LRS
2) User identification strategy for LTI launch
3) Data model changes to tenants, users, courses tables
4) UI fixes for privacy dashboard
5) highchart, highchart-ng upgrates to 5.0.11 and 1.0.1
6) Donut chart for privacy dashboard
7) Added gulp build to minify static assets
8) package.json and bower.json updates.
9) Ebextension updates - gulp build, unset iframe deny options in apache configs
10) Updates to ingestion function to validate and accept learning statements appropriately.

### CODE REVIEW

File: node_modules/lrs-core/lib/api.js [hunk 1]
```diff
@@ -132,3 +151,15 @@ var initializeAuthorizationMiddleware = function(apiRouter) {
...
+/**
+ * Function to generate a random integer in a given range
+ *
+ * @param  {Int}       low          Lower limit in the range
+ * @param  {Int}       high         Higher limit in the range
+ * @param {Int}        randomInt    Random integer
+ */
+var randomInt = function randomInt (low, high) {
+    return Math.floor(Math.random() * (high - low) + low);
+}
```
Comment (pauline2k) [inline node_modules/lrs-core/lib/api.js]: @sandeepmjay What if we used `_.random` here instead of rolling our own? https://lodash.com/docs/4.17.4#random

Comment (pauline2k) [reply node_modules/lrs-core/lib/api.js]: @sandeepmjay Actually, looks like this function is unused.

File: node_modules/lrs-course/lib/api.js [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+  // Get the course from the DB or create it if it doesn't exist yet
+  options = {
+    'where': {
+      'canvas_course_id': canvasCourseId
+    },
+    'defaults': {
+      'tenant_api_domain': canvas.tenant_api_domain,
+      'name': courseInfo.name,
+      'privacydashboard_url': courseInfo.privacydashboard_url
+
```
Comment (pauline2k) [inline node_modules/lrs-course/lib/api.js]: @sandeepmjay Stray whitespace

File: node_modules/lrs-lti/lib/api.js [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        // Create the user on the fly
+        var defaults = {
+          'canvas_course_role': canvas_course_role,
+          'canvas_full_name': canvas_full_name,
+          'canvas_image': canvas_image,
+          'canvas_email': canvas_email,
+          'canvas_enrollment_state': canvas_enrollment_state,
+          'external_id': canvas_user_login_id.toString(),
+          'tenant_id': canvas.id
+
```
Comment (pauline2k) [inline node_modules/lrs-lti/lib/api.js]: @sandeepmjay Stray whitespace

File: node_modules/lrs-lti/lib/api.js [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+        UsersAPI.getOrCreateUser(canvas_user_id, course, defaults, function(err, user) {
+          if (err) {
+            return callback(err);
+          }
+          log.info('Found or created user :' + JSON.stringify(user));
+          // Keep track of the URL that is performing the LTI launch
+          provider.body.tool_url = req.headers.referer;
+
+          log.info('Provider Tool URL' + provider.body.tool_url)
```
Comment (pauline2k) [inline node_modules/lrs-lti/lib/api.js]: @sandeepmjay Add colon and space before URL?

File: node_modules/lrs-lti/lib/constants.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+ * SOFTWARE AND ACCOMPANYING DOCUMENTATION, IF ANY, PROVIDED HEREUNDER IS PROVIDED
+ * "AS IS". REGENTS HAS NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, UPDATES,
+ * ENHANCEMENTS, OR MODIFICATIONS.
+ */
+
+var Constants = module.exports = {
+  'PRIVACYDASHBOARD': {
+    'id': 'privacydashboard',
+    'title': 'Student Privacy Dashboard',
+    'description': 'Privacy Dashboard provides students transparency on how their Learning data is being used.'
```
Comment (pauline2k) [inline node_modules/lrs-lti/lib/constants.js]: @sandeepmjay Does capitalized `Learning` look odd?

File: node_modules/lrs-statements/lib/caliper.js [hunk 6]
```diff
@@ -166,37 +167,63 @@ var getOrCreateUser = function(ctx, statement, callback) {
...
+  var canvas_image = null;
+  var canvas_full_name = null;
+  var canvas_enrollment_state = null;
 
   // TODO : Add support for other actor types like Organizations based on examples from Caliper spec.
   if (statement.actor.hasOwnProperty('type') && statement.actor.hasOwnProperty('id') && statement.actor.type === 'Person') {
-    externalId = statement.actor.id.split('/').pop();
-    console.log("External ID for the user is: " + externalId);
+    canvas_global_user_id = statement.actor.id.split('/').pop();
+    if(statement.actor.hasOwnProperty('extensions')) {
```
Comment (pauline2k) [inline node_modules/lrs-statements/lib/caliper.js]: @sandeepmjay Space after `if`

File: node_modules/lrs-statements/lib/caliper.js [hunk 7]
```diff
@@ -166,37 +167,63 @@ var getOrCreateUser = function(ctx, statement, callback) {
...
+      external_id = statement.actor.extensions[0].user_login;
+    }
+
   } else if (statement.actor.hasOwnProperty('type') && statement.actor.type === 'SoftwareApplication') {
     try {
       if (statement.object.actor.hasOwnProperty('id') && statement.object.actor.hasOwnProperty('type') && statement.object.actor.type === 'Person') {
-        externalId = statement.object.actor.id.split('/').pop();
-        console.log("External ID for the user is: " + externalId);
+        canvas_global_user_id = statement.object.actor.id.split('/').pop();
+        if(statement.object.actor.hasOwnProperty('extensions')) {
```
Comment (pauline2k) [inline node_modules/lrs-statements/lib/caliper.js]: @sandeepmjay Space after if

File: node_modules/lrs-course/lib/api.js (hunk 2 above)
Comment (sandeep-jay) [reply node_modules/lrs-course/lib/api.js]: Fixed

File: public/app/privacydashboard/data/datasources.json [hunk 8]
```diff
@@ -1,10 +1,19 @@
...
   "name": "bCourses",
-  "total": 5421
+  "total": 3421
 }, {
-  "name": "CalCentral",
-  "total": 3458
+  "name": "Piazza",
+  "total": 700
+}, {
+  "name": "Turnit In",
```
Comment (pauline2k) [inline public/app/privacydashboard/data/datasources.json]: @sandeepmjay I think `Turnitin` is all one word.

File: public/app/privacydashboard/datasources/datasourcesController.js [hunk 9]
```diff
@@ -23,125 +23,137 @@
...
+     /**
+      * Get the data sources donut chart
+      *
+      * @api private
+      */
+     var getChart = function() {
+       var chartIndex = $('#privacydashboard-datasources-chart').data('highcharts-chart');
+       var chart = Highcharts.charts[chartIndex];
+       return chart.reflow();
+
```
Comment (pauline2k) [inline public/app/privacydashboard/datasources/datasourcesController.js]: @sandeepmjay Stray whitespace

File: node_modules/lrs-lti/lib/api.js (hunk 4 above)
Comment (sandeep-jay) [reply node_modules/lrs-lti/lib/api.js]: That was for debugging. Removing this.

File: node_modules/lrs-lti/lib/api.js (hunk 3 above)
Comment (sandeep-jay) [reply node_modules/lrs-lti/lib/api.js]: Fixed

File: public/app/privacydashboard/data/datasources.json (hunk 8 above)
Comment (sandeep-jay) [reply public/app/privacydashboard/data/datasources.json]: It is

**Comment by pauline2k** [comment]
Thank you, @sandeepmjay; I have nothing but picayune comments on this. Good to merge once relinted.

File: node_modules/lrs-core/lib/api.js (hunk 1 above)
Comment (sandeep-jay) [reply node_modules/lrs-core/lib/api.js]: Removed this.

**Comment by sandeep-jay** [comment]
@paulkerschen, Had to look up picayune. Stop making me work hard :)! Relinting completed. Go ahead with the merge.

**Pull request merged**